File: hello-tts-rust/src/config_manager.rs
use crate::tts_client::{TTSConfig, TTSError};
use log::info;
use std::collections::HashMap;
use std::path::Path;

/// Configuration manager with preset support
pub struct ConfigManager;

impl ConfigManager {
    const DEFAULT_CONFIG_PATHS: &'static [&'static str] =
        &["./tts_config.json", "~/.tts/config.json"];

    /// Get predefined presets
    pub fn get_presets() -> HashMap<&'static str, TTSConfig> {
        let mut presets = HashMap::new();

        presets.insert("default", TTSConfig::default());

        presets.insert(
            "fast",
            TTSConfig {
                rate: "+20%".to_string(),
                max_concurrent: 5,
                batch_size: 10,
                ..TTSConfig::default()
            },
        );

        presets.insert(
            "slow",
            TTSConfig {
                rate: "-20%".to_string(),
                max_concurrent: 2,
                batch_size: 3,
                ..TTSConfig::default()
            },
        );

        presets.insert(
            "high_quality",
            TTSConfig {
                output_format: "wav".to_string(),
                cache_voices: true,
                max_retries: 5,
                ..TTSConfig::default()
            },
        );

        presets.insert(
            "batch_processing",
            TTSConfig {
                max_concurrent: 8,
                batch_size: 20,
                cache_voices: true,
                ..TTSConfig::default()
            },
        );

        presets.insert(
            "whisper",
            TTSConfig {
                rate: "-10%".to_string(),
                volume: "50%".to_string(),
                pitch: "-5%".to_string(),
                ..TTSConfig::default()
            },
        );

        presets.insert(
            "excited",
            TTSConfig {
                rate: "+15%".to_string(),
                pitch: "+10%".to_string(),
                volume: "110%".to_string(),
                ..TTSConfig::default()
            },
        );

        presets
    }

    /// Load configuration from file or use default
    pub fn load_config(config_path: Option<&str>) -> Result<TTSConfig, TTSError> {
        if let Some(path) = config_path {
            return TTSConfig::from_json_file(path);
        }

        // Try default paths
        for path in Self::DEFAULT_CONFIG_PATHS {
            let expanded_path = Self::expand_path(path);
            if Path::new(&expanded_path).exists() {
                return TTSConfig::from_json_file(&expanded_path);
            }
        }

        // Return default config if no file found
        Ok(TTSConfig::default())
    }

    /// Get a preset configuration
    pub fn get_preset(preset_name: &str) -> Result<TTSConfig, TTSError> {
        let presets = Self::get_presets();
        presets.get(preset_name).cloned().ok_or_else(|| {
            let available: Vec<_> = presets.keys().collect();
            let available_str = available
                .iter()
                .map(|s| s.to_string())
                .collect::<Vec<_>>()
                .join(", ");
            TTSError::Config(format!(
                "Unknown preset '{}'. Available: {}",
                preset_name, available_str
            ))
        })
    }

    /// List available preset names
    pub fn list_presets() -> Vec<&'static str> {
        Self::get_presets().keys().cloned().collect()
    }

    /// Create a default configuration file
    pub fn create_default_config(file_path: &str, preset: &str) -> Result<(), TTSError> {
        let config = Self::get_preset(preset)?;
        config.to_json_file(file_path)?;
    info!("Created default configuration file: {}", file_path);
        Ok(())
    }

    /// Expand path with home directory
    fn expand_path(path: &str) -> String {
        if path.starts_with("~/") {
            if let Some(home) = std::env::var_os("HOME") {
                return path.replace("~", &home.to_string_lossy());
            }
        }
        path.to_string()
    }
}

/// Convenience functions
pub fn load_config(config_path: Option<&str>) -> Result<TTSConfig, TTSError> {
    ConfigManager::load_config(config_path)
}

pub fn get_preset(preset_name: &str) -> Result<TTSConfig, TTSError> {
    ConfigManager::get_preset(preset_name)
}

pub fn create_default_config(file_path: &str, preset: &str) -> Result<(), TTSError> {
    ConfigManager::create_default_config(file_path, preset)
}

pub fn list_presets() -> Vec<&'static str> {
    ConfigManager::list_presets()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_config() {
        let config = TTSConfig::default();
        assert_eq!(config.default_voice, "en-US-AriaNeural");
        assert_eq!(config.output_format, "mp3");
        assert_eq!(config.batch_size, 5);
        assert_eq!(config.max_concurrent, 3);
    }

    #[test]
    fn test_presets() {
        let presets = ConfigManager::get_presets();
        assert!(presets.contains_key("default"));
        assert!(presets.contains_key("fast"));
        assert!(presets.contains_key("slow"));

        let fast_config = presets.get("fast").unwrap();
        assert_eq!(fast_config.rate, "+20%");
        assert_eq!(fast_config.max_concurrent, 5);
    }

    #[test]
    fn test_get_preset() {
        let config = ConfigManager::get_preset("fast").unwrap();
        assert_eq!(config.rate, "+20%");
        assert_eq!(config.batch_size, 10);
    }

    #[test]
    fn test_unknown_preset() {
        let result = ConfigManager::get_preset("unknown");
        assert!(result.is_err());
    }

    #[test]
    fn test_list_presets() {
        let presets = ConfigManager::list_presets();
        assert!(presets.contains(&"default"));
        assert!(presets.contains(&"fast"));
        assert!(presets.contains(&"slow"));
    }

    #[test]
    fn test_config_validation() {
        let mut config = TTSConfig::default();
        assert!(config.validate().is_ok());

        config.default_voice = String::new();
        assert!(config.validate().is_err());

        config.default_voice = "en-US-AriaNeural".to_string();
        config.batch_size = 0;
        assert!(config.validate().is_err());
    }
}

File: hello-tts-java/src/main/java/org/feuyeux/tts/config/ConfigManager.java
package org.feuyeux.tts.config;

import lombok.extern.slf4j.Slf4j;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Configuration manager with preset support
 */
@Slf4j
public class ConfigManager {
    private static final String[] DEFAULT_CONFIG_PATHS = {
        "./tts_config.json",
        System.getProperty("user.home") + "/.tts/config.json"
    };

    private static final Map<String, TTSConfig> PRESETS = new HashMap<>();

    static {
        // Initialize presets
        PRESETS.put("default", new TTSConfig());
        
        PRESETS.put("fast", new TTSConfig.Builder()
            .rate("+20%")
            .maxConcurrent(5)
            .batchSize(10)
            .build());
        
        PRESETS.put("slow", new TTSConfig.Builder()
            .rate("-20%")
            .maxConcurrent(2)
            .batchSize(3)
            .build());
        
        PRESETS.put("high_quality", new TTSConfig.Builder()
            .outputFormat("wav")
            .cacheVoices(true)
            .maxRetries(5)
            .build());
        
        PRESETS.put("batch_processing", new TTSConfig.Builder()
            .maxConcurrent(8)
            .batchSize(20)
            .cacheVoices(true)
            .build());
        
        PRESETS.put("whisper", new TTSConfig.Builder()
            .rate("-10%")
            .volume("50%")
            .pitch("-5%")
            .build());
        
        PRESETS.put("excited", new TTSConfig.Builder()
            .rate("+15%")
            .pitch("+10%")
            .volume("110%")
            .build());
    }

    /**
     * Load configuration from file or use default
     */
    public static TTSConfig loadConfig(String configPath) throws IOException {
        if (configPath != null) {
            return loadFromFile(configPath);
        }

        // Try default paths
        for (String path : DEFAULT_CONFIG_PATHS) {
            if (Files.exists(Paths.get(path))) {
                return loadFromFile(path);
            }
        }

        // Return default config if no file found
        return new TTSConfig();
    }

    /**
     * Load configuration from specific file
     */
    public static TTSConfig loadFromFile(String filePath) throws IOException {
        Path path = Paths.get(filePath);
        if (!Files.exists(path)) {
            throw new FileNotFoundException("Configuration file not found: " + filePath);
        }

        String content = Files.readString(path);
        
        // Simple JSON parsing (in a real implementation, you'd use a JSON library)
        Map<String, Object> configMap = parseSimpleJson(content);
        return TTSConfig.fromMap(configMap);
    }

    /**
     * Save configuration to file
     */
    public static void saveConfig(TTSConfig config, String filePath) throws IOException {
        Path path = Paths.get(filePath);
        Files.createDirectories(path.getParent());

        String json = mapToJson(config.toMap());
        Files.writeString(path, json);
    }

    /**
     * Get a preset configuration
     */
    public static TTSConfig getPreset(String presetName) {
        if (!PRESETS.containsKey(presetName)) {
            String available = String.join(", ", PRESETS.keySet());
            throw new IllegalArgumentException("Unknown preset '" + presetName + "'. Available: " + available);
        }
        return PRESETS.get(presetName);
    }

    /**
     * List available preset names
     */
    public static Set<String> listPresets() {
        return PRESETS.keySet();
    }

    /**
     * Create a default configuration file
     */
    public static void createDefaultConfig(String filePath, String preset) throws IOException {
        TTSConfig config = getPreset(preset);
        saveConfig(config, filePath);
        log.info("Created default configuration file: {}", filePath);
    }

    /**
     * Simple JSON parser (for demonstration - use a proper JSON library in production)
     */
    private static Map<String, Object> parseSimpleJson(String json) {
        Map<String, Object> result = new HashMap<>();
        
        // Remove braces and split by comma
        json = json.trim().replaceAll("^\\{|}$", "");
        String[] pairs = json.split(",");
        
        for (String pair : pairs) {
            String[] keyValue = pair.split(":", 2);
            if (keyValue.length == 2) {
                String key = keyValue[0].trim().replaceAll("\"", "");
                String value = keyValue[1].trim();
                // Parse different types
                Object parsedValue;
                if (value.equals("true")) {
                    parsedValue = true;
                } else if (value.equals("false")) {
                    parsedValue = false;
                } else if (value.startsWith("\"") && value.endsWith("\"")) {
                    parsedValue = value.substring(1, value.length() - 1);
                } else {
                    try {
                        parsedValue = Integer.parseInt(value);
                    } catch (NumberFormatException e) {
                        parsedValue = value;
                    }
                }
                
                result.put(key, parsedValue);
            }
        }
        
        return result;
    }

    /**
     * Simple JSON serializer (for demonstration - use a proper JSON library in production)
     */
    private static String mapToJson(Map<String, Object> map) {
        StringBuilder json = new StringBuilder("{\n");
        
        boolean first = true;
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (!first) {
                json.append(",\n");
            }
            first = false;
            
            json.append("  \"").append(entry.getKey()).append("\": ");
            
            Object value = entry.getValue();
            if (value instanceof String) {
                json.append("\"").append(value).append("\"");
            } else {
                json.append(value);
            }
        }
        
        json.append("\n}");
        return json.toString();
    }
}

